Add AD7091R-2/-4/-8 driver and project

machschmitt opened the pull request:
Initial implementation of the ad7091r8 no-OS driver. This adds support for ad7091r-2/-4/-8 devices which roughly support the same operations as ad7091r5, but are interfaced through SPI.

Driver has now been tested with Maxim78000FTHR and ad7091r8-sdz evaluation board.

Other meaningful differences between ad7091r8 and ad7091r5 are:
- ad7091r8 has a normal mode which comprises ad7091r5's sample and command modes
- ad7091r8 has no autocycle mode
- ad7091r8 and ad7091r5 have slightly different register fields and contrast most on their configuration register
- ad7091r8 protocol requires a pulse in the CONVST pin and so the driver requests a GPIO for that during initialization

amiclaus commented on drivers/adc/ad7091r8/ad7091r8.c:
```diff
@@ -0,0 +1,566 @@
+ */
+uint16_t ad7091r8_pulse_convst(struct ad7091r8_dev *dev)
+{
+	int32_t ret = 0;
```
remove explicit initialization

amiclaus commented on drivers/adc/ad7091r8/ad7091r8.c:
```diff
@@ -0,0 +1,566 @@
+	int32_t ret = 0;
+
+	/* convst pulse width must be 10 ns minimum, 500 ns maximum */
+	ret |= no_os_gpio_set_value(dev->gpio_convst, NO_OS_GPIO_LOW);
```
check ret after each operation (avoid stacking error codes with `|=`)

amiclaus commented on drivers/adc/ad7091r8/ad7091r8.c:
```diff
@@ -0,0 +1,566 @@
+		return -EINVAL;
+
+	ret = ad7091r8_pulse_convst(dev);
+	if (ret < 0)
```
`(ret)`

amiclaus commented on drivers/adc/ad7091r8/ad7091r8.c:
```diff
@@ -0,0 +1,566 @@
+
+	/* Assumes controller host machine is little-endian */
+	buf[0] = AD7091R8_REG_WRITE(reg_addr);
+	buf[0] |= reg_data & 0x03;
```
do the `buff[0]` in one line. also what does 0x3 mean? add suggestive macro.

amiclaus commented on drivers/adc/ad7091r8/ad7091r8.c:
```diff
@@ -0,0 +1,566 @@
+		return -EINVAL;
+
+	ret = ad7091r8_pulse_convst(dev);
+	if (ret < 0)
```
ditto

amiclaus commented on drivers/adc/ad7091r8/ad7091r8.c:
```diff
@@ -0,0 +1,566 @@
+	/* Big-endian to little-endian */
+	*reg_data = (buf[0] << 8) | buf[1];
+
+	return ret;
```
return 0

amiclaus commented on drivers/adc/ad7091r8/ad7091r8.c:
```diff
@@ -0,0 +1,566 @@
+		return ret;
+
+	/* Big-endian to little-endian */
+	*reg_data = (buf[0] << 8) | buf[1];
```
use please provided functions like `no_os_get_unaligned_be16`

amiclaus commented on drivers/adc/ad7091r8/ad7091r8.c:
```diff
@@ -0,0 +1,566 @@
+		return -EINVAL;
+
+	ret = ad7091r8_spi_reg_read(dev, reg_addr, &reg_data);
+	if (ret < 0)
```
ditto.

amiclaus commented on drivers/adc/ad7091r8/ad7091r8.h:
```diff
@@ -0,0 +1,261 @@
+
+/* AD7091R8_REG_CONF */
+#define REG_CONF_SLEEP_MODE_MASK	(NO_OS_BIT(1) | NO_OS_BIT(0))
+#define REG_CONF_SLEEP_MODE(x)		((x & 0x03) << 0)
```
remove all these defines with variable. use `no_os_field_get` / `no_os_field_prep` inline.

amiclaus commented on drivers/adc/ad7091r8/ad7091r8.c:
```diff
@@ -0,0 +1,566 @@
+		break;
+	default:
+		return -EINVAL;
+		break;
```
no need for `break;` here

amiclaus commented on drivers/adc/ad7091r8/ad7091r8.c:
```diff
@@ -0,0 +1,566 @@
+	switch (limit) {
+	case AD7091R8_LOW_LIMIT:
+		reg = AD7091R8_REG_CH_LOW_LIMIT(channel);
+		value += (value & 0x04) ? 8 : 0;
```
avoid if possible ternary operators.

amiclaus commented on drivers/adc/ad7091r8/ad7091r8.c:
```diff
@@ -0,0 +1,566 @@
+
+	ret = no_os_spi_init(&dev->spi_desc, &init_param.spi_init);
+	if (ret < 0)
+		return ret;
```
please use conditional jumps and handle all the memory leaks in this function.

amiclaus commented on drivers/adc/ad7091r8/ad7091r8.c:
```diff
@@ -0,0 +1,566 @@
+		return ret;
+
+	no_os_gpio_get_optional(&dev->gpio_reset, init_param.gpio_reset);
+	if (dev->gpio_reset != NULL) {
```
just `if (dev->gpio_reset)`

amiclaus commented on drivers/adc/ad7091r8/ad7091r8.c:
```diff
@@ -0,0 +1,566 @@
+	if (dev->gpio_reset != NULL) {
+		ret = no_os_gpio_direction_output(&dev->gpio_reset,
+						  NO_OS_GPIO_HIGH);
+		if (ret < 0)
```
ditto

amiclaus commented on drivers/adc/ad7091r8/ad7091r8.c:
```diff
@@ -0,0 +1,566 @@
+		ad7091r8_reset(dev, false);
+	}
+
+	no_os_gpio_get_optional(&dev->gpio_alert, init_param.gpio_alert);
```
handle error code.

amiclaus commented on drivers/adc/ad7091r8/ad7091r8.c:
```diff
@@ -0,0 +1,566 @@
+ * 		      - false: hardware reset
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int32_t ad7091r8_reset(struct ad7091r8_dev *dev, bool is_software)
```
general aspect. for return types of the functions we use standard `int`.

amiclaus commented on drivers/adc/ad7091r8/ad7091r8.c:
```diff
@@ -0,0 +1,566 @@
+	if (ret < 0)
+		return ret;
+
+	dev->gpio_reset = NULL;
```
are these explicit initializations mandatory?

amiclaus commented on drivers/adc/ad7091r8/ad7091r8.c:
```diff
@@ -0,0 +1,566 @@
+	dev->device_id = init_param.device_id;
+
+	ret = no_os_gpio_get(&dev->gpio_convst, init_param.gpio_convst);
+	if (ret < 0)
```
actually check the entire code for `(ret < 0)` -> `(ret)`

amiclaus commented on drivers/adc/ad7091r8/ad7091r8.c:
```diff
@@ -0,0 +1,566 @@
+
+	*device = dev;
+
+	return ret;
```
return 0

amiclaus reviewed (commented):
some initial comments on my side

machschmitt commented:
All suggestions from previous review have been applied.

Change log v1 -> v2:

- Made use of no_os_put_unaligned_be16() and similar functions to avoid unaligned access.
- Made use of no_os_field_prep(), no_os_field_prep() and bit masks to access register fileds in a more readable way.
- Function return types changed form int32_t to int.
- Changed return value checks: if (ret < 0) -> if (ret)
- Several fixes and improvements to ad7091r8 driver
- Added iio_ad7091r8 driver for IIO interface support
- Added project for using AD7091R-8 with Maxim78000FTHR containing a dummy example and a simple IIO example.

buha commented on drivers/adc/ad7091r8/ad7091r8.c:
```diff
@@ -0,0 +1,609 @@
+int ad7091r8_set_sleep_mode(struct ad7091r8_dev *dev,
+			    enum ad7091r8_sleep_mode mode)
+{
+	if (!dev)
```
this check is useless here, dev is not used in this function, it's only passed to ad7091r8_spi_write_mask which will anyway return error if dev is NULL

buha commented on drivers/adc/ad7091r8/ad7091r8.c:
```diff
@@ -0,0 +1,609 @@
+	uint16_t mask;
+	uint16_t val;
+
+	if (!dev)
```
same remark here, you can skip the dev check

buha commented on drivers/adc/ad7091r8/ad7091r8.c:
```diff
@@ -0,0 +1,609 @@
+{
+	uint16_t value;
+
+	if (!dev)
```
...and here and probably throughout the whole file

buha commented on drivers/adc/ad7091r8/iio_ad7091r8.c:
```diff
@@ -0,0 +1,268 @@
+/******************************************************************************/
+/************************ Functions Declarations ******************************/
+/******************************************************************************/
+//extern int ad7091r8_read_device_data(struct ad7091r8_dev *dev,
```
what about these commented declarations here ?

buha commented on drivers/adc/ad7091r8/iio_ad7091r8.c:
```diff
@@ -0,0 +1,268 @@
+/************************ Functions Definitions *******************************/
+/******************************************************************************/
+
+//extern int ad7091r8_read_device_data(struct ad7091r8_dev *dev,
```
do something about these commented declarations

buha commented on drivers/adc/ad7091r8/iio_ad7091r8.c:
```diff
@@ -0,0 +1,268 @@
+	desc->iio_dev = &ad7091r8_iio_dev;
+
+	// Initialize ad7091r8 driver
+	//ret = ad7091r8_init(&desc->ad7091r8_dev, *(init_param->ad7091r8_dev_init));
```
delete this line

buha commented on drivers/adc/ad7091r8/iio_ad7091r8.h:
```diff
@@ -0,0 +1,71 @@
+struct ad7091r8_iio_dev {
+	struct ad7091r8_dev *ad7091r8_dev;
+	struct iio_device *iio_dev;
+	//int ad7091r8_hpf_3db_table[7][2];
```
yeah so plenty of such commented out sections, please delete them all

danmois commented on drivers/adc/ad7091r8/ad7091r8.c:
```diff
@@ -0,0 +1,609 @@
+	case AD7091R8_LOW_LIMIT:
+		reg = AD7091R8_REG_CH_LOW_LIMIT(channel);
+		/* Approximate reg value to requested limit value */
+		if (value & 0x04)
```
Defines could be used here.

buha commented on projects/ad7091r8-sdz/builds.json:
```diff
@@ -0,0 +1,10 @@
+{
+    "maxim": {
+        "dummy_example_max78000_ad7091r8": {
```
It looks like this project is tailored to run on max78000 and this means you're connecting the eval board to the max78000 with flywires. In this case it's not easy to figure out what the connections should be so I think it's necessary to add a documentation .rst file as well with a table showing the required HW connections.

danmois commented on drivers/adc/ad7091r8/ad7091r8.c:
```diff
@@ -0,0 +1,609 @@
+		break;
+	case AD7091R8_HIGH_LIMIT:
+		reg = AD7091R8_REG_CH_HIGH_LIMIT(channel);
+		*value = 0x07;
```
Please check all the code for places where magic numbers can be replaced by defines.

danmois commented on drivers/adc/ad7091r8/ad7091r8.c:
```diff
@@ -0,0 +1,609 @@
+}
+
+/***************************************************************************//**
+ * @brief Initializes the communication peripheral and the initial Values for
```
Inconsistent documentation format.

danmois commented on drivers/adc/ad7091r8/ad7091r8.c:
```diff
@@ -0,0 +1,609 @@
+	if (!device)
+		return -EINVAL;
+
+	dev = (struct ad7091r8_dev *)no_os_malloc(sizeof(*dev));
```
no_os_calloc would be a better choice here.

danmois commented on drivers/adc/ad7091r8/ad7091r8.c:
```diff
@@ -0,0 +1,609 @@
+	if (ret)
+		printf("WARNING: Failed to RESET on power up\n\r");
+
+	//ret = no_os_gpio_get_optional(&dev->gpio_alert, init_param->gpio_alert);
```
Should this commented code remain in the driver?

danmois commented on drivers/adc/ad7091r8/ad7091r8.c:
```diff
@@ -0,0 +1,609 @@
+	if (ret)
+		return ret;
+
+	ret = ad7091r8_spi_reg_read(dev, AD7091R8_REG_RESULT, read_val);
```
You can return directly.

danmois commented on drivers/adc/ad7091r8/ad7091r8.h:
```diff
@@ -0,0 +1,269 @@
+/***************************************************************************//**
+ *   @file   ad7091r8.h
+ *   @brief  Implementation of AD7091R-8 Driver
```
Header file.

danmois commented on drivers/adc/ad7091r8/iio_ad7091r8.c:
```diff
@@ -0,0 +1,268 @@
+/******************************************************************************/
+/************************ Functions Declarations ******************************/
+/******************************************************************************/
+//extern int ad7091r8_read_device_data(struct ad7091r8_dev *dev,
```
Should this commented block of code remain in the driver?

danmois replied:
Valid for other places too.

danmois commented on projects/ad7091r8-sdz/src/examples/dummy/dummy_example.c:
```diff
@@ -0,0 +1,90 @@
+/******************************************************************************/
+/***************************** Include Files **********************************/
+/******************************************************************************/
+#include "dummy_example.h"
```
Maybe the naming is not the most fortunate one. basic_example ?

danmois commented on projects/ad7091r8-sdz/src/platform/maxim/main.c:
```diff
@@ -0,0 +1,107 @@
+	if (NO_OS_IS_ERR_VALUE(ret))
+		goto error;
+
+	//Debug TODO remove
```
Should this be removed?

danmois commented on projects/ad7091r8-sdz/src/platform/maxim/parameters.h:
```diff
@@ -0,0 +1,88 @@
+#define SPI_EXTRA       &ad7091r8_spi_extra_ip
+
+#define GPIO_OPS        &max_gpio_ops
+#define GPIO_EXTRA      &ad7091r8_gpio_extra_ip
```
Alignment would be beneficial for readability reasons.

danmois reviewed (commented):
Looks good. A little cleanup work is needed.

machschmitt commented:
Applied all suggestions to v2.
Added a new example for faster sampling rates with triggered buffer support.
Updated the wiki page https://wiki.analog.com/resources/tools-software/uc-drivers/ad7091r8

Change log v2 -> v3:

- Dropped approximation for limit register values.
- Added define for high limit register LSBs magic number.
- Removed unneeded NULL checks.
- Made documentation more uniform
- Used calloc for device struct memory allocation
- Removed old unnecessary comments
- Directly returned from ad7091r8_read_one() after SPI reg read.
- Specified if file is a header file in file comments.
- Renamed DUMMY_EXAMPLE to BASIC_EXAMPLE
- Removed debug print from maxim main.c
- Added tabs to get better alignment in parameters.h
- Added build targets for for AD7091R-2 and AD7091R-4.
- Added a delay to single-shot readings to get correct ADC output codes.
- Added timer trigger support for supporting faster sampling rates.
- Added project README file with documentation and setup instructions.

buha commented on drivers/adc/ad7091r8/ad7091r8.c:
```diff
@@ -0,0 +1,578 @@
+
+	ret = ad7091r8_reset(dev, !dev->gpio_reset);
+	if (ret)
+		printf("WARNING: Failed to RESET on power up\n\r");
```
I would remove the printf statements and return some error code to the caller instead. This one in particular is wrong anyway in the sense that it's a very bad thing not to be able to reset the chip and still continue (no return) configuring. The chip is in an unknown state if not reset.

danmois commented on drivers/adc/ad7091r8/ad7091r8.c:
```diff
@@ -0,0 +1,578 @@
+
+	ret = no_os_gpio_remove(dev->gpio_reset);
+	if (ret)
+		printf("Failed to realease RESET GPIO");
```
Functions from no_os_print_log.h would be more appropriate for printing messages (e.g., pr_info, pr_err, ...).

machschmitt commented:
Change log v3 -> v4:
- Replaced all prints in ad7091r8.c by error code return or jump to the proper error handling path.

rbolboac commented on drivers/adc/ad7091r8/iio_ad7091r8.c:
```diff
@@ -0,0 +1,364 @@
+	if (!desc)
+		return -ENOMEM;
+
+	if (!init_param || !init_param->ad7091r8_dev_init) {
+		ret = -EINVAL;
+		goto error_ad7091r8_init;
+	}
```
these can be checked before the memory allocation above

rbolboac commented on drivers/adc/ad7091r8/iio_ad7091r8.h:
```diff
@@ -0,0 +1,81 @@
+/******************************************************************************/
+/************************ Functions Declarations ******************************/
+/******************************************************************************/
+int ad7091r8_iio_init(struct ad7091r8_iio_dev **iio_dev,
+		      struct ad7091r8_iio_dev_init_param *init_param);
+
+int ad7091r8_iio_remove(struct ad7091r8_iio_dev *desc);
```
missing APIs doxygen comments

rbolboac reviewed (commented):
some comment from my side

machschmitt commented on drivers/adc/ad7091r8/ad7091r8.h:
```diff
@@ -0,0 +1,273 @@
+	AD7091R8,
+};
+
+static char * const ad7091r8_names[] = {
+	[AD7091R2] = "ad7091r-2",
+	[AD7091R4] = "ad7091r-4",
+	[AD7091R8] = "ad7091r-8",
+};
```
ad7091r8_names[] is accessed with ad7091r8_init_p->device_id in projects to get the specific device name.

machschmitt commented:
Hi everyone, thanks for all the suggestions and tips provided for this PR.
Updated version with most review suggestions applied.
Comments on points I've not changed.

Change log v4 -> v5:
- Updated source files copyright from 2023 to 2024.
- Removed check on zero value write.
- Removed << 3 from limit register read data. This provides raw data from the
  device and is coherent with the way the linux driver outputs limit register values.
- Removed unneeded else clause and put code right after preceding if.
- Fixed error handling gotos/jumps in ad7091r8_init().
- Avoided pointer cast with local variable for register read.
- Brought param check to begin of ad7091r8_iio_init().
- Added doxygen doc comments to iio_ad7091r8.h
- Removed delay specific to MAX78000 from base ad7091r8 driver.

rbolboac commented on drivers/adc/ad7091r8/iio_ad7091r8.c:
```diff
@@ -0,0 +1,372 @@
+	struct ad7091r8_iio_dev *desc;
+
+	if (!init_param || !init_param->ad7091r8_dev_init) {
+		ret = -EINVAL;
```
return directly, desc is not allocated yet

rbolboac commented on projects/ad7091r8-sdz/Makefile:
```diff
@@ -0,0 +1,15 @@
+# For instance, build with `make BASIC_EXAMPLE=y` to choose the BASIC_EXAMPLE.
+
+# Force select an example by assigning y for enabling.
+#BASIC_EXAMPLE = y
```
maybe also add an IIO_EXAMPLE = y line, if both are supported by the project

rbolboac commented on projects/ad7091r8-sdz/src/platform/maxim/main.c:
```diff
@@ -0,0 +1,100 @@
+	ret = basic_example_main();
+
+error:
+	no_os_uart_remove(uart_desc);
```
this is wrong.... if uart_init fails, no_os_uart_remove does not have to be called, simply return ret in this case

rbolboac commented on projects/ad7091r8-sdz/src/examples/iio_example/iio_example.c:
```diff
@@ -0,0 +1,92 @@
+
+	ret = iio_app_init(&app, app_init_param);
+	if (ret)
+		return ret;
```
ad7091r8_iio_remove should be called in case of an error. you may take a look at the handling here, and follow the same logic: https://github.com/analogdevicesinc/no-OS/blob/main/projects/max14916/src/examples/iio_example/iio_example.c

rbolboac commented on projects/ad7091r8-sdz/src/examples/iio_timer_trigger_example/iio_timer_trigger_example.c:
```diff
@@ -0,0 +1,151 @@
+
+	ret = iio_app_init(&app, app_init_param);
+	if (ret)
+		return ret;
```
please fix the memory leaks in this function in case of an error...make sure you call all remove APIs if the init API has been successful

rbolboac commented on projects/ad7091r8-sdz/src/examples/basic/basic_example.c:
```diff
@@ -0,0 +1,90 @@
+
+error:
+	printf("Error on AD7091R-8 basic_example: %d\n\r", ret);
+	return ret;
```
call ad7091r8_remove before returning

amiclaus commented on drivers/adc/ad7091r8/ad7091r8.c:
```diff
@@ -0,0 +1,558 @@
+	uint16_t value;
+
+	switch (mode) {
+	case AD7091R8_GPO0_ENABLED:
```
hmm, what's the point of doing this switch case when you have the `enum ad7091r8_gpo0_mode mode` and you can initialized the reg values directly there:
```
/**
 * @enum ad7091r8_gpo0_mode
 * @brief Port 0 configuration
 */
enum ad7091r8_gpo0_mode {
	/** GPO0 is output port */
	AD7091R8_GPO0_ENABLED = 0,
	/** GPO0 is Alert indicator */
	AD7091R8_GPO0_ALERT = 16,
	/** GPO0 is busy indicator, device is converting */
	AD7091R8_GPO0_BUSY = 48,
};
```

machschmitt replied:
Nice. thanks!

amiclaus commented on drivers/adc/ad7091r8/ad7091r8.c:
```diff
@@ -0,0 +1,558 @@
+		return ret;
+
+	no_os_free(dev);
+	return ret;
```
nit: return 0

amiclaus commented on drivers/adc/ad7091r8/ad7091r8.h:
```diff
@@ -0,0 +1,268 @@
+
+#ifndef __AD7091R8_H__
+#define __AD7091R8_H__
+
```
drop extra line

amiclaus commented on drivers/adc/ad7091r8/iio_ad7091r8.c:
```diff
@@ -0,0 +1,372 @@
+	/* Dummy read cycle to let the conversion sequence get updated */
+	return ad7091r8_sequenced_read(ad7091r8, &dummy);
+}
+
```
nit: drop extra line

amiclaus commented on drivers/adc/ad7091r8/iio_ad7091r8.c:
```diff
@@ -0,0 +1,372 @@
+
+	return iio_buffer_push_scan(iio_dev_data->buffer, data_buff);
+}
+
```
ditto

amiclaus reviewed (commented):
minor comments on my side.

machschmitt commented:
Change log v5 -> v6:
- Fixed error handling code. No free of unallocated structs.
- Added IIO_EXAMPLE line to project Makefile.
- Added list of supported devices in project Makefile.
- In project basic example, return directly if no_os_uart_init() fails.
- Fixed error handling code in project examples and avoided mem leaks by calling proper resource remove functions.
- Simplified ad7091r8_gpo0_mode as suggested
- Minor cleanups as suggested.

rbolboac reviewed (approved):
lgtm
